### PR TITLE
[PATCH v4] linux-gen: ipsec: add configurable ordering mode

### PR DESCRIPTION
**Pull request by JannePeltonen**
Add new config file parameters to control how original input queue
order is maintained during asynchronous IPsec processing.

When asynchronous IPsec operations are started within an ordered
scheduling context, packet order is maintained in IPsec completion
queues but the order in which the sequence number gets assigned or
the anti-replay check done is not guaranteed. This commit adds a config
option to force the stateful part of IPsec processing to be done in the
correct order, at the cost of reduced parallelism.

The ordering mechanism added in this commit uses the order_lock method
of the internal scheduling API to stop the calling thread until it
is its turn (i.e the preceding scheduling contexts associated with the
same input queue have been released). Future commits may add more
elaborate mechanisms.

The default behaviour does not change. By default no ordering is done.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>

### CODE REVIEW

**Comment by JannePeltonen** [comment]
v2: Added config file version bumps that were missing from the fist version.

File: config/odp-linux-generic.conf [hunk 1]
```diff
@@ -234,3 +234,51 @@ timer: {
...
+	#    Lack of ordering means that outbound IPsec packets, although
+	#    remaining in the correct order, may have their sequence numbers
+	#    assigned out of order. This can cause unexpected packet loss if
+	#    the anti-replay window of the receiving end is not large enough
+	#    to cover the possible misordering.
+	#
+	#    Similarly, since anti-replay check is not done in the reception
+	#    order, the anti-replay check sees additional packet misordering
+	#    on top of the true misordering of the received packets. This a
+	#    larger anti-replay window is required to avoid packet loss.
```
Comment (JereLeppanen) [inline config/odp-linux-generic.conf:267]: Last sentence needs fixing (grammar).

**Comment by JannePeltonen** [comment]
v3: Fixed the botched comment sentence in the config file.

**Review by JereLeppanen — APPROVED** [review]

**Comment by JannePeltonen** [comment]
v4: rebased and added review-tag

**Pull request merged**